cli/cmd/edges.go

```diff
@@ -0,0 +1,297 @@
+					}
+
+					rows := edgesRespToRows(resp)
+
+					c <- indexedEdgeResults{num, rows, err}
+				}(num, req)
+			}
+
+			totalRows := make([]*pb.Edge, 0)
+			i := 0
+			for res := range c {
+				if res.err != nil {
+					return res.err
+				}
+				totalRows = append(totalRows, res.rows...)
+				if i++; i == len(reqs) {
+					close(c)
+				}
+			}
+
+			output := renderEdgeStats(totalRows, options)
+			_, err = fmt.Print(output)
+
+			return err
+		},
+	}
+
+	cmd.PersistentFlags().StringVarP(&options.namespace, "namespace", "n", options.namespace, "Namespace of the specified resource")
+	cmd.PersistentFlags().StringVarP(&options.outputFormat, "output", "o", options.outputFormat, "Output format; one of: \"table\" or \"json\"")
+	return cmd
+}
+
+func buildEdgesRequests(resources []string, options *statOptions) ([]*pb.EdgesRequest, error) {
+	targets, err := util.BuildResources(options.namespace, resources)
+	if err != nil {
+		return nil, err
+	}
+
+	requests := make([]*pb.EdgesRequest, 0)
+	for _, target := range targets {
+		err = options.validate(target.Type)
+		if err != nil {
+			return nil, err
+		}
+
+		requestParams := util.EdgesRequestParams{
+			ResourceType: target.Type,
+			Namespace:    options.namespace,
+		}
+
+		req, err := util.BuildEdgesRequest(requestParams)
+		if err != nil {
+			return nil, err
+		}
+		requests = append(requests, req)
+	}
+	return requests, nil
+}
+
+func edgesRespToRows(resp *pb.EdgesResponse) []*pb.Edge {
+	rows := make([]*pb.Edge, 0)
+	if resp != nil {
+		for _, edgeTable := range resp.GetOk().Edges {
+			rows = append(rows, edgeTable)
+		}
+	}
+	return rows
+}
+
+func requestEdgesFromAPI(client pb.ApiClient, req *pb.EdgesRequest) (*pb.EdgesResponse, error) {
+	resp, err := client.Edges(context.Background(), req)
+	if err != nil {
+		return nil, fmt.Errorf("Edges API error: %+v", err)
+	}
+	if e := resp.GetError(); e != nil {
+		return nil, fmt.Errorf("Edges API response error: %+v", e.Error)
+	}
+	return resp, nil
+}
+
+func renderEdgeStats(rows []*pb.Edge, options *statOptions) string {
+	var buffer bytes.Buffer
+	w := tabwriter.NewWriter(&buffer, 0, 0, padding, ' ', tabwriter.AlignRight)
+	writeEdgesToBuffer(rows, w, options)
+	w.Flush()
+
+	return renderEdges(buffer, &options.statOptionsBase)
+}
+
+type edgeRowStats struct {
+	src    string
+	dst    string
+	client string
+	server string
+	msg    string
+}
+
+func writeEdgesToBuffer(rows []*pb.Edge, w *tabwriter.Writer, options *statOptions) {
+	edgeTables := make(map[string]*edgeRowStats)
+	if len(rows) != 0 {
+		for _, r := range rows {
+			key := string(r.Dst.Name + r.Src.Name)
+
+			edgeTables[key] = &edgeRowStats{
+				client: r.ClientId,
+				server: r.ServerId,
+				msg:    r.NoIdentityMsg,
+				src:    r.Src.Name,
+				dst:    r.Dst.Name,
+			}
+		}
+	}
+	switch options.outputFormat {
+	case tableOutput:
+		if len(edgeTables) == 0 {
+			fmt.Fprintln(os.Stderr, "No edges found.")
+			os.Exit(0)
+		}
+		printSingleEdgeTable(edgeTables, w)
```

Comment:
i think we either want to explicitly support multiple resource types (similar to `stat`), or disallow it:
```bash
$ bin/linkerd edges -n linkerd deploy/ po/
SRC                                                                   DST                                                                     CLIENT                                                                     SERVER   MSG
linkerd-controller-86ddf787bb-wz44v   linkerd-prometheus-595b874c57-zx8nw   linkerd-controller.linkerd.serviceaccount.identity.linkerd.cluster.local   linkerd-prometheus.linkerd.serviceaccount.identity.linkerd.cluster.local     -
linkerd-controller                                     linkerd-prometheus   linkerd-controller.linkerd.serviceaccount.identity.linkerd.cluster.local   linkerd-prometheus.linkerd.serviceaccount.identity.linkerd.cluster.local     -
```

```bash
$ bin/linkerd stat -n linkerd deploy/ po/
NAME                                         MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
deploy/linkerd-controller                       1/1   100.00%   0.9rps           1ms           8ms          10ms         10
deploy/linkerd-grafana                          1/1   100.00%   0.3rps           1ms           2ms           2ms          2
...

NAME                                         MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
po/linkerd-controller-86ddf787bb-wz44v          1/1   100.00%   0.9rps           1ms           8ms          10ms         10
po/linkerd-grafana-c46ccbbd7-4r2x7              1/1   100.00%   0.3rps           1ms           2ms           2ms          2
...
```

Reply:
Thanks for catching that. Multiple resource types seems like something that would be 👍 to support and would take just a few tweaks.

Reply:
On second thought I'd like to merge a simpler command that only accepts one resource type, then go from there. I changed `newCmdEdges` to accept only one resource argument with `cobra. ExactArgs(1)`.

Reply:
@scottcarol If you're only accepting one resource type for now, then you might wanna validate against having `all` being passed, as it's currently returning something like this:
```
$ bin/go-run cli edges -n emojivoto all
SRC   DST              CLIENT          SERVER   MSG
            default.emojivoto   web.emojivoto     -
```

Reply:
@alpeb I believe I've addressed this in my latest commit. Thanks for the thorough tests of this feature! 🙏